src/tests.rs

```diff
@@ -0,0 +1,258 @@
+
+    // ReadOnly and WriteOnly
+    assert_eq!(
+        unsafe { VolatilePtr::new_with_access(NonNull::from(&mut val), Access::read_only()) }
+            .read(),
+        51
+    );
+    unsafe { VolatilePtr::new_with_access(NonNull::from(&mut val), Access::write_only()) }
+        .write(12);
+    assert_eq!(val, 12);
+
+    // Custom: safe read + safe write
+    {
+        let access = Access {
+            read: SafeAccess,
+            write: SafeAccess,
+        };
+        let mut volatile = unsafe { VolatilePtr::new_with_access(NonNull::from(&mut val), access) };
+        let random: i32 = rand::random();
+        volatile.write(i64::from(random));
+        assert_eq!(volatile.read(), i64::from(random));
+        let random2: i32 = rand::random();
+        volatile.update(|i| *i += i64::from(random2));
+        assert_eq!(volatile.read(), i64::from(random) + i64::from(random2));
+    }
+
+    // Custom: safe read + unsafe write
+    {
+        let access = Access {
+            read: SafeAccess,
+            write: UnsafeAccess,
+        };
+        let mut volatile = unsafe { VolatilePtr::new_with_access(NonNull::from(&mut val), access) };
+        let random: i32 = rand::random();
+        unsafe { volatile.write_unsafe(i64::from(random)) };
+        assert_eq!(volatile.read(), i64::from(random));
+        let random2: i32 = rand::random();
+        unsafe { volatile.update_unsafe(|i| *i += i64::from(random2)) };
+        assert_eq!(volatile.read(), i64::from(random) + i64::from(random2));
+    }
+
+    // Custom: safe read + no write
+    {
+        let access = Access {
+            read: SafeAccess,
+            write: NoAccess,
+        };
+        let random = rand::random();
+        val = random;
+        let volatile = unsafe { VolatilePtr::new_with_access(NonNull::from(&mut val), access) };
+        assert_eq!(volatile.read(), i64::from(random));
+    }
+
+    // Custom: unsafe read + safe write
+    {
+        let access = Access {
+            read: UnsafeAccess,
+            write: SafeAccess,
+        };
+        let mut volatile = unsafe { VolatilePtr::new_with_access(NonNull::from(&mut val), access) };
+        let random: i32 = rand::random();
+        volatile.write(i64::from(random));
+        assert_eq!(unsafe { volatile.read_unsafe() }, i64::from(random));
+        let random2: i32 = rand::random();
+        unsafe { volatile.update_unsafe(|i| *i += i64::from(random2)) };
+        assert_eq!(
+            unsafe { volatile.read_unsafe() },
+            i64::from(random) + i64::from(random2)
+        );
+    }
+
+    // Todo: Custom: unsafe read + unsafe write
+    // Todo: Custom: unsafe read + no write
+    // Todo: Custom: no read + safe write
+    // Todo: Custom: no read + unsafe write
+    // Todo: Custom: no read + no write
+
+    // Todo: is there a way to check that a compile error occurs when trying to use
+    // unavailable methods (e.g. `write` when write permission is `UnsafeAccess`)?
```

Comment:
I think you can have a failing doctest? That might be the easiest way to do it.

See: https://doc.rust-lang.org/edition-guide/rust-2018/rustdoc/documentation-tests-can-now-compile-fail.html